python-package/xgboost/collective.py

```diff
@@ -0,0 +1,217 @@
+"""XGBoost collective communication related API."""
+import ctypes
+from enum import IntEnum, unique
+import logging
+import pickle
+from typing import Any, TypeVar, Optional, cast, List, Union
+
+import numpy as np
+
+from .core import _LIB, c_str, _check_call
+
+LOGGER = logging.getLogger("[xgboost.collective]")
+
+
+def _init_collective() -> None:
+    """internal library initializer."""
+    if _LIB is not None:
+        _LIB.XGCommunicatorGetRank.restype = ctypes.c_int
+        _LIB.XGCommunicatorGetWorldSize.restype = ctypes.c_int
+        _LIB.XGCommunicatorIsDistributed.restype = ctypes.c_int
+
+
+def init(args: Optional[List[bytes]] = None) -> None:
+    """Initialize the collective library with arguments"""
+    if args is None:
+        args = []
+    arr = (ctypes.c_char_p * len(args))()
+    arr[:] = cast(List[Union[ctypes.c_char_p, bytes, None, int]], args)
+    _LIB.XGCommunicatorInit(len(arr), arr)
+
+
+def finalize() -> None:
+    """Finalize the process, notify tracker everything is done."""
+    _LIB.XGCommunicatorFinalize()
+
+
+def get_rank() -> int:
+    """Get rank of current process.
+
+    Returns
+    -------
+    rank : int
+        Rank of current process.
+    """
+    ret = _LIB.XGCommunicatorGetRank()
+    return ret
+
+
+def get_world_size() -> int:
+    """Get total number workers.
+
+    Returns
+    -------
+    n : int
+        Total number of process.
+    """
+    ret = _LIB.XGCommunicatorGetWorldSize()
+    return ret
+
+
+def is_distributed() -> int:
+    """If the collective communicator is distributed."""
+    is_dist = _LIB.XGCommunicatorIsDistributed()
+    return is_dist
+
+
+def communicator_print(msg: Any) -> None:
+    """Print message to the communicator.
+
+    This function can be used to communicate the information of
+    the progress to the communicator.
+
+    Parameters
+    ----------
+    msg : str
+        The message to be printed to the communicator.
+    """
+    if not isinstance(msg, str):
+        msg = str(msg)
+    is_dist = _LIB.XGCommunicatorIsDistributed()
+    if is_dist != 0:
+        _check_call(_LIB.XGCommunicatorPrint(c_str(msg)))
+    else:
+        print(msg.strip(), flush=True)
+
+
+def get_processor_name() -> bytes:
+    """Get the processor name.
+
+    Returns
+    -------
+    name : str
+        the name of processor(host)
+    """
+    mxlen = 256
+    length = ctypes.c_ulong()
+    buf = ctypes.create_string_buffer(mxlen)
```

Comment:
XGBoost has a thread-local store for returning buffer. Take a look at the C API on how various strings and floating point buffers are returned.